Artillery ability implementation

Badulundabad opened the pull request:
- added ArtilleryDeploying ability super class;
- adjusted ability logic;
- added tor_other_item.xml and added mortar item there;
- removed checking for IsAbilityUser in CustomCrosshairMissionBehavior because it's obious if you have ability component and any ability.

For campaign:
You need to purchase artillery (this option has not implemented yet, so for now only via cheat). In mission caster is limited by 1 piece per 50 points of engineering skill;
For custom battle:
Amount of artillery is hardcoded. Currently only empire lord can use it with limit of 5 pieces.

Badulundabad committed:
ordered the ability templates by types

Badulundabad committed:
added ArtilleryDeploying super ability type

Badulundabad committed:
adjusted abillity logic

Badulundabad committed:
Merge branch 'development' into artillery_rework

Badulundabad committed:
removed debug stuff

Badulundabad committed:
made some changes

Badulundabad committed:
removed excess IsAbilityUser checks

Badulundabad committed:
renamed ArtilleryDeploying class

hunharibo commented on CSharpSourceCode/Abilities/AbilityManagerMissionLogic.cs:
```diff
@@ -201,9 +201,11 @@ public override void OnAgentCreated(Agent agent)
             {
-                if (agent.IsAbilityUser())
+                var character = agent.Character as CharacterObject;
+                bool canHasArtillery = (Game.Current.GameType is Campaign) && character != null && character.IsHero && character.HeroObject.IsPartyLeader;
```
This should be driven by a new character attribute instead of being hardcoded to partyleader heroes. We might want engineer companions or special NPCs with this ability as well. Create a "CanPlaceArtillery" character attribute string and check for that with character.HasAttribute("CanPlaceArtillery") instead.

Badulundabad replied:
Ok. I've created Agent extension CanPlaceArtillery instead and moved this checking to AbilityComponent

hunharibo commented on CSharpSourceCode/Battle/CrosshairMissionBehavior/CustomCrosshairMissionBehavior.cs:
```diff
@@ -100,7 +100,7 @@ private void InitializeCrosshairs()
             _weaponCrosshair.InitializeCrosshair();
             _sniperScope = new SniperScope();
 
-            if (Agent.Main.IsAbilityUser() && (_abilityComponent = Agent.Main.GetComponent<AbilityComponent>()) != null)
```
Is there any particulat use case for removing this as well? Or just to avoid a double check. I am always hesitant to remove double checks, they don't bother anyone and who knows what it might protect against.

Badulundabad replied:
Yeah, sometimes it's better to check twice, but there we have more. Here are all places we had or still have IsAbilityUser checks:
CustomCrosshairMissionBehavior:
1) CustomCrosshairMissionBehavior.OnMissionScreenTick() - to check if ability crosshair should be shown;
2) AbilityManagerMissionLogic.HandleInput() - to check if player agent should run its extension CastCurrentAbility(), but HandleInput method also checks if current ability crosshair is visible what mean
3) Ability.CanCast() aslo to check if caster can use ability;

Badulundabad replied:
i think here we can remove this one. It's nothing to worry about if crosshairs are initialized, but player is not an ability user, because to see crosshair you anyway pass check during OnMissionScreenTick()

hunharibo replied:
as discussed, agreed. resolving

hunharibo reviewed (changes requested):
Very nice work. Few comments and one request for the new attribute. Add the attribute to the main player in charactercontentcreation as well if the picked last option is engineer.
Also make the artillery "item" a merchandize, because we want to buy/sell these. Test please if that is possible with cheat menu.

Badulundabad commented on CSharpSourceCode/Abilities/AbilityComponent.cs:
```diff
@@ -11,8 +14,30 @@ namespace TOW_Core.Abilities
-        public AbilityComponent(Agent agent) : base(agent)
+        public AbilityComponent(Agent agent, bool canHasArtillery) : base(agent)
         {
+            if (canHasArtillery)
```
no i'm not, thanks😄

Badulundabad replied:
fixed

Badulundabad commented on CSharpSourceCode/Abilities/Ability.cs:
```diff
@@ -87,7 +87,7 @@ public virtual bool CanCast(Agent casterAgent)
             return !_isCasting &&
                    !IsOnCooldown() &&
                    ((casterAgent.IsPlayerControlled && (Crosshair.CrosshairType == CrosshairType.CenteredAOE || IsRightAngleToCast())) || 
-                   (casterAgent.IsActive() && casterAgent.Health > 0 && casterAgent.GetMorale() > 1 && casterAgent.IsAbilityUser()));
```
you're right, i will get it back

Badulundabad replied:
fixed

Badulundabad committed:
changed ability stuff regaring hunharibo suggestions

- got back IsAbilityUser checking to Ability and SpecialMove classes just to prevent illegal using of abilities;
- created an Agent extension CanPlaceArtillery to check agent for required attribute;
- changed AbilityComponent constructor and AbilityManagerMissionLogic regarding CanPlaceArtillery  extension.

Badulundabad committed:
returned back IsAbilityUser check

Badulundabad committed:
added CanPlaceArtillery attribute to empire lord

Badulundabad committed:
Merge branch 'development' into artillery_rework

Z3rca reviewed (commented):
looks overall good, does it also work for the AI?

Badulundabad commented:
I think no, it will be another pr

Badulundabad committed:
added Place Great Cannon ability

hunharibo commented:
Merged manually
